op-node/rollup/derive/batch_queue.go

```diff
@@ -4,205 +4,291 @@ import (
+		return ret
 	}
 
-	bq.inputs = bq.inputs[1:]
-
-	return attributes, nil
 }
 
-func (bq *BatchQueue) Step(ctx context.Context, outer Progress) error {
-	if changed, err := bq.progress.Update(outer); err != nil {
-		return err
-	} else if changed {
-		if !bq.progress.Closed { // init inputs if we moved to a new open origin
-			bq.inputs = append(bq.inputs, BatchesWithOrigin{Origin: bq.progress.Origin, Batches: nil})
-		}
+// tryPopNextBatch tries to get the next batch from the batch queue using an eager approach.
+func (bq *BatchQueue) tryPopNextBatch(l2SafeHead eth.L2BlockRef) *BatchWithL1InclusionBlock {
+	// We require at least 1 L1 blocks to look at.
+	if len(bq.l1Blocks) == 0 {
 		return nil
 	}
-
-	attrs, err := bq.DeriveL2Inputs(ctx, bq.next.SafeL2Head().Time)
-	if err != nil {
-		return err
+	batches, ok := bq.batchesByTimestamp[l2SafeHead.Time+bq.config.BlockTime]
+	// No more batches found.
+	if !ok {
+		return nil
 	}
-	for _, attr := range attrs {
-		if uint64(attr.Timestamp) <= bq.next.SafeL2Head().Time {
-			// drop attributes if we are still progressing towards the next stage
-			// (after a reset rolled us back a full sequence window)
-			continue
+
+	// Find the first batch saved for this timestamp.
+	// Note that we expect the number of batches for the same timestamp to be small (frequently just 1 ).
+	for _, batch := range batches {
+		l1OriginTime := bq.l1Blocks[0].Time
+
+		// If this batch advances the epoch, check it's validity against the next L1 Origin
+		if batch.Batch.EpochNum != rollup.Epoch(l2SafeHead.L1Origin.Number) {
+			// With only 1 l1Block we cannot look at the next L1 Origin.
+			// Note: This means that we are unable to determine validity of a batch
+			// without more information. In this case we should bail out until we have
+			// more information otherwise the eager algorithm may diverge from a non-eager
+			// algorithm.
+			if len(bq.l1Blocks) < 2 {
+				bq.log.Warn("eager batch wants to advance epoch, but could not")
+				return nil
+			}
+			l1OriginTime = bq.l1Blocks[1].Time
 		}
-		bq.log.Info("derived new payload attributes", "time", uint64(attr.Timestamp), "txs", len(attr.Transactions))
-		bq.next.AddSafeAttributes(attr)
-	}
-	return nil
-}
 
-func (bq *BatchQueue) ResetStep(ctx context.Context, l1Fetcher L1Fetcher) error {
-	// if we only just started resetting, find the origin corresponding to the safe L2 head
-	if !bq.resetting {
-		l2SafeHead := bq.next.SafeL2Head()
-		l1SafeHead, err := l1Fetcher.L1BlockRefByHash(ctx, l2SafeHead.L1Origin.Hash)
+		// Timestamp bounds
+		minL2Time := l2SafeHead.Time + bq.config.BlockTime
+		maxL2Time := l1OriginTime + bq.config.MaxSequencerDrift
+		if minL2Time+bq.config.BlockTime > maxL2Time {
+			maxL2Time = minL2Time + bq.config.BlockTime
+		}
+
+		// Note: Don't check epoch change here, check it in `validExtension`
+		epoch, err := bq.dl.L1BlockRefByNumber(context.TODO(), uint64(batch.Batch.EpochNum))
```

Comment:
`ctx` should be passed in, and the function should return `*BatchWithL1InclusionBlock, error` to handle the error. Same for the `deriveBatches` func which calls this func. And the error should be handled as non-critical in the `Step` func, i.e. a failed RPC call should not result in an EOF or error, just a nil + subsequent retry.